### ISSUE TITLE
[Issue] Creating Duplicate Services on FluentValidator Startup Registration

### ISSUE THREAD

**Issue by pdevito3** [FluentValidation/FluentValidation#1334]
### System Details

- FluentValidation version: aspnetcore & DependencyInjectionExtensions 8.6.1
- Web Framework version: dotnet core 3.1

### Issue Description
So I have the attached startup and validator files which, on the surface, work fine when running my API:

[FV_PetStartup.txt](https://github.com/FluentValidation/FluentValidation/files/4251136/FV_PetStartup.txt)
[FV_PetValidators.txt](https://github.com/FluentValidation/FluentValidation/files/4251135/FV_PetValidators.txt)

### The Issue
My issue is, when I run a simple integration test like this:
````
        [Fact]
        public async Task PostPetReturnsSuccessCodeAndResourceWithAccurateFields()
        {
            // Arrange
            var client = _factory.CreateClient();
            var fakePet = new FakePetDto().Generate();

            // Act
            var httpResponse = await client.PostAsJsonAsync("/v1/Pets", fakePet)
                .ConfigureAwait(false);

            // Assert
            httpResponse.EnsureSuccessStatusCode();

            var resultDto = JsonConvert.DeserializeObject<PetDto>(await httpResponse.Content.ReadAsStringAsync()
                .ConfigureAwait(false));

            httpResponse.StatusCode.Should().Be(201);
            response.Should().ContainEquivalentOf(fakePet);
        }
````

It is throwing this error:
> System.InvalidOperationException: 'Sequence contains more than one matching element'

### My Findings So Far
I know fluent validators is the issue because I can comment out my validators registration in startup and the integration test doesn't throw the error anymore.

I dug into it and, as far as I can tell, the configureservices method in my startup is creating duplicate validator services, which would make sense given the error message. I've tried numerous iterations of fluent validator registrations on addmvc:
````
//.AddFluentValidation(fv =>
            //{
            //    //fv.RegisterValidatorsFromAssemblyContaining<PetForManipulationDtoValidator>();
            //    //fv.RegisterValidatorsFromAssemblyContaining<PetForUpdateDtoValidator>();
            //    fv.RegisterValidatorsFromAssemblyContaining<PetForCreationDtoValidator>();
            //});
            //.AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblies(AppDomain.CurrentDomain.GetAssemblies()));
            //RegisterValidatorsFromAssemblyContaining<StartupDevelopment>());
````

None of which have worked, even a simple breakdown with just one abstract validator. Hoping some fresh eyes can see what I'm missing.

**Comment by JeremySkinner** [comment]
Can you post the full stacktrace for the exception? Thanks

**Comment by JeremySkinner** [comment]
Yep, that'll be it. Looks AddFluentValidation is being called twice, once directly and once by Autofac's DelegateActivator. I'd suggest putting together a small sample project that reproduces the issue as the next step.

**Comment by pdevito3** [comment]
Roger. New to autofac, so good to know! Noted on next steps. Will be sure to follow up here

**Comment by JeremySkinner** [comment]
If you also wanted to post your autofac setup here, then I might be able to spot the issue. If you're doing any kind of automatic wiring for all types, make sure you're not duplicating the registrations that FluentValidation will be performing directly.

**Comment by pdevito3** [comment]
Appreciate it.

Program.cs is quite barebones:
````
public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .UseServiceProviderFactory(new AutofacServiceProviderFactory())
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup(typeof(Startup).GetTypeInfo().Assembly.FullName)
                    .UseContentRoot(Directory.GetCurrentDirectory())
                    .UseKestrel()
                    .UseNLog();
                });
````

And nothing crazy in my ConfigureContainer block
````
public void ConfigureContainer(ContainerBuilder builder)
        {
            builder.Register(c => new MapperConfiguration(cfg =>
            {
                cfg.AddProfile(new PetProfile(c.Resolve<ILogger<PetProfile>>()));
            }).CreateMapper());

            builder.RegisterType<PetRepository>().As<IPetRepository>();
            builder.RegisterType<SieveProcessor>();
        }
````

**Comment by JeremySkinner** [comment]
What's in the PetProfile?

**Comment by pdevito3** [comment]
Just an automapper profile:
````
namespace mnglabs.pets.api.Configuration
{
    using System.Collections.Generic;
    using AutoMapper;
    using Microsoft.AspNetCore.JsonPatch;
    using Microsoft.AspNetCore.JsonPatch.Operations;
    using Microsoft.Extensions.Logging;
    using mnglabs.pets.api.Data.Entities;
    using mnglabs.pets.Data;
    using Newtonsoft.Json;

    public class PetProfile : Profile
    {
        public PetProfile(ILogger<PetProfile> logger)
        {
            //createmap<to this, from this>
            CreateMap<Pet, PetDto>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<PetForCreationDto, Pet>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<PetForUpdateDto, Pet>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<Fret, FretForUpdateDto>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<Fret,FretDto>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<CrPet, CrPetDto>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();
            CreateMap<CrPet, CrPetForUpdateDto>()
                .BeforeMap((src, dest) => LogMap(logger, src, dest))
                .AfterMap((src, dest) => LogMap(logger, src, dest))
                .ReverseMap();

            CreateMap(typeof(JsonPatchDocument<>), typeof(JsonPatchDocument<>));
            CreateMap(typeof(Operation<>), typeof(Operation<>));
        }

        public static void LogMap(ILogger logger, object src, object dest)
        {
            logger.LogInformation($"Mapping Source: {JsonConvert.SerializeObject(src)} to Destination: {JsonConvert.SerializeObject(dest)}");
        }
    }
}
````

**Comment by JeremySkinner** [comment]
Ah ok, thought it might be something Autofac related. In that case I'd suggest continuing to put together a minimal project to reproduce the problem, then if you want to put it on github somewhere I can try and take a look, but might not be until later in the week.

**Comment by pdevito3** [comment]
Sounds good. Appreciate the responsiveness

**Comment by pdevito3** [comment]
@JeremySkinner so I'm not quite sure what was happening before, but I was able to get it working.

I can use this to register the validators:
````
            .AddFluentValidation(fv => fv.RegisterValidatorsFromAssembly(Assembly.GetExecutingAssembly()));
````

And if I add this block in autofac, it resolves the issue for whatever reason.
````
            builder.RegisterAssemblyTypes(typeof(Startup).GetType().Assembly)
                .Where(t => t.IsClosedTypeOf(typeof(IValidator<>)))
                .AsImplementedInterfaces();
````

If you can shed some light on how FV interacts with autofac and what the recommended intergartion is, I'd love to hear, otherwise, Just happy it's working. Thanks again!

**Comment by JeremySkinner** [comment]
FV doesn’t interact with autofocus directly. As part of AddFluentValidation it loops over the assemblies you specified and finds any validator classes and adds them to the asp.net service provider (in your case, you’ve swapped out the normal service provider with an autofac version, which is where autofac comes in).

 It also adds the FluentValidationModelValidatorProvider to the asp.net model validator providers list *this is the bit that’s the problem* as it should only ever be added once, and in your application it’s being added twice (and why I think there’s something in your application calling AddFluentValidation a second time). Here’s the code in question: https://github.com/FluentValidation/FluentValidation/blob/master/src/FluentValidation.AspNetCore/FluentValidationMvcExtensions.cs#L66

**Comment by JeremySkinner** [comment]
Oh and one other thing that'd be worth doing: Put a breakpoint in your ConfigureServices method and check that it's not being called twice.

**Comment by pdevito3** [comment]
Agreed. Still not sure what's causing the multiple calls though. Breakpoint on configure services only triggers once.

Best I can tell, the startup is configured correctly, but it has something to do with the way the WebHostbuilder is using 'UseStartup' when it specifies the assembly containing the startup like this:
````
webBuilder.UseStartup(typeof(Startup).GetTypeInfo().Assembly.FullName)
````

Because when I specify the startup directly like this, it works fine.
````
webBuilder.UseStartup<StartupDevelopment>()
````

Maybe the former is somehow trigger it twice when it gets the assembly that the breakpoint doesn't pick up?

Regardless, I made a [barebones project with weatherforecast](https://github.com/pdevito3/FvAutofac) to demonstrate it if you want to check it out. As is, it will use the 'broken' startup, but you can toggle them. **The integration test will trigger the issue (or not depending on which startup you use)**
````
                    //webBuilder.UseStartup<Startup>()
                    webBuilder.UseStartup(typeof(Startup).GetTypeInfo().Assembly.FullName)
````


Wish I could go with the explicit startup, but it isn't really feasible given that I need multiple startups for my environments.

**Comment by JeremySkinner** [comment]
Ok, thanks. I'm planning to spend a couple of hours working on FluentValidation on Friday, so I'll try and dig into this then.

**Comment by pdevito3** [comment]
Sorry to spam you, but FYI for Friday, I don't think it's autofac related at all. If I remove all of the autofac code, it will still throw the same error if using the the 'broken' startup listed above :-/

I pushed a branch with autofac removed for you to check out instead if you want. Thanks again!

**Comment by JeremySkinner** [comment]
I'm looking at this today, I'll post notes as I find them

**Comment by JeremySkinner** [comment]
1) When using `webBuilder.UseStartup(typeof(Startup).GetTypeInfo().Assembly.FullName)` , `Startup.ConfigureServices` is indeed being called twice, so we can be sure this isn't related to FluentValidation

**Comment by JeremySkinner** [comment]
Ok, I've spent a few hours on this and I've not really been able to get any further than where you did. Understanding the complex interactions between WebApplicationFactory and the WebHost factory is probably better left up to the asp.net team to troubleshoot. But we can definitely be certain the double startup isn't related to FluentValidation or any other 3rd party component. 

> I'm not sure why FV FluentValidationModelValidatorProvider is the only thing that gets duplicated

This is because the FV provider is not being registered as a service. The services registrations are being overwritten/recreated when ConfigureServices runs the second time, but the FVModelValidatorProvider is being added to the `ModelValidatorProviders` collection as part of a call to `AddMvcOptions` (see https://github.com/FluentValidation/FluentValidation/blob/master/src/FluentValidation.AspNetCore/FluentValidationMvcExtensions.cs#L66-L80) Because the `MvcBuilder` *is* the same instance across both calls to ConfigureServices, the configuration logic ends up being registered twice (and run twice). 

What I've done is add a check if the provider has been added to the collection already and skip registering again (7776bff1383299a404b4a612d84fe399f92cb21b). We shouldn't really have to do this, but works around the problem until MS fix it in core. This will be part of 9.0 (I'll try and push out a 9.0-preview3 either later today or tomorrow) 

Anyway, at least you have a workaround for now.
